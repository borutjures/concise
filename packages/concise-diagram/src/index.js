// @flow

/* eslint-disable prefer-template */

import fs from 'fs';
import Viz from 'viz.js';
import { addDefaults } from 'timm';
import type {
  Schema,
  OutputProcessor,
  SchemaUtils,
  ModelName,
  FieldName,
} from 'concise-types';

type OutputOptions = {
  file?: string,
  filterEdges?: (
    { from: ModelName, to: ModelName, as: FieldName, required: boolean },
  ) => boolean,
  edgeLabels?: boolean,
};

const DEFAULT_OPTIONS = {
  edgeLabels: true,
};

// ====================================
// Main
// ====================================
const output: OutputProcessor = async (
  schema: Schema,
  options: OutputOptions,
  utils: SchemaUtils,
) => {
  const preprocessedSchema = utils.preprocess(schema);
  const vizInput = writeDiagram(
    preprocessedSchema,
    addDefaults(options, DEFAULT_OPTIONS),
  );
  const svg = Viz(vizInput);
  if (options.file) {
    fs.writeFileSync(options.file, svg, 'utf8');
  }
  return svg;
};

// ====================================
// Flow writer
// ====================================
const writeDiagram = ({ models }, { filterEdges, edgeLabels }) => {
  const modelNames = Object.keys(models);
  const nodes = modelNames.map(modelName => {
    let node = modelName;
    const { description } = models[modelName];
    const props = [];
    if (description) {
      props.push(`comment="${description}"`);
      props.push(`tooltip="${description}"`);
    }
    if (props.length) node += ` [${props.join(', ')}]`;
    return node;
  });
  const edges = [];
  modelNames.forEach(modelName => {
    const { relations } = models[modelName];
    Object.keys(relations).forEach(fieldName => {
      const specs = relations[fieldName];
      if (specs.isInverse) return;
      const required = specs.validations && specs.validations.required === true;
      if (
        filterEdges &&
        !filterEdges({
          from: modelName,
          to: specs.model,
          as: fieldName,
          required,
        })
      ) {
        return;
      }
      const props = [];
      if (edgeLabels && fieldName !== specs.model) {
        props.push(`label=${fieldName}`);
      }
      if (!required) props.push('style=dotted');
      let edge = `${modelName} -> ${specs.model}`;
      if (props.length) edge += ` [${props.join(', ')}]`;
      edges.push(edge);
    });
  });
  return 'digraph "" {\n' +  // `""` removes the background tooltip
    '  node [shape=box, fontname="sans-serif"];\n' +
    '  edge [fontsize=9, fontname="sans-serif"];\n' +
    nodes.map(o => `  ${o};\n`).join('') +
    edges.map(o => `  ${o};\n`).join('') +
    '}\n';
};

// ====================================
// Public
// ====================================
export { output };
