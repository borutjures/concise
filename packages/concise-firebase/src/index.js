// @flow

/* eslint-disable prefer-template, no-param-reassign */

import fs from 'fs';
import type { Schema, OutputProcessor, SchemaUtils } from 'concise-types';

/* --
Output-only.

Output options:
* `file?` (`string`): if specified, output will be written to the specified path
-- */
type OutputOptions = {
  file?: string,
};

const VALIDATE = '.validate';
const ISO_8601 =
  '/^\\d\\d\\d\\d-\\d\\d-\\d\\dT\\d\\d:\\d\\d:\\d\\d(\\.\\d+)?(([+-]\\d\\d:\\d\\d)|Z)?$/i';

// ====================================
// Main
// ====================================
const output: OutputProcessor = async (
  schema: Schema,
  options: OutputOptions,
  utils: SchemaUtils,
) => {
  const rules = writeRulesForAllModels(utils.preprocessedSchema);
  if (options.file) {
    fs.writeFileSync(options.file, JSON.stringify(rules, null, 2), 'utf8');
  }
  return rules;
};

// ====================================
// Rule writer
// ====================================
const writeRulesForAllModels = ({ models }) => {
  const rules = {};
  Object.keys(models).forEach(modelName => {
    if (!models[modelName].existsInServer) return;
    rules[modelName] = {
      [`$${modelName}Id`]: writeRulesForModel(models, modelName),
    };
  });
  return rules;
};

const writeRulesForModel = (models, modelName) => {
  const model = models[modelName];
  const modelRules = {};

  // Required fields
  const requiredFields = getRequiredFields(model);
  if (requiredFields.length) {
    const list = requiredFields.map(o => `'${o}'`).join(', ');
    modelRules[VALIDATE] = `newData.hasChildren([${list}])`;
  }

  // Field rules
  writeRulesForModelFields(modelRules, model);

  return modelRules;
};

const writeRulesForModelFields = (modelRules, { fields, relations }) => {
  const rules = {};
  Object.keys(fields).forEach(name => {
    const field = fields[name];
    if (field.isPrimaryKey) return;
    const constraints = getFieldConstraints(field);
    if (constraints.length) {
      if (modelRules[name] == null) modelRules[name] = {};
      modelRules[name][VALIDATE] = constraints.join(' && ');
    }
  });
  Object.keys(relations).forEach(name => {
    const relation = relations[name];
    const { fkName, isPlural } = relation;
    if (modelRules[fkName] == null) modelRules[fkName] = {};
    const fieldRules = modelRules[fkName];
    if (isPlural) {
      if (fieldRules.$id == null) fieldRules.$id = {};
      fieldRules.$id[VALIDATE] = 'newData.val() === true';
    } else {
      fieldRules[VALIDATE] = 'newData.isString()';
    }
  });
  return rules;
};

const getFieldConstraints = field => {
  const constraints = [];

  // Type rules
  const { type } = field;
  if (type === 'string' || type === 'uuid') {
    constraints.push('newData.isString()');
  } else if (type === 'boolean') {
    constraints.push('newData.isBoolean()');
  } else if (type === 'number') {
    constraints.push('newData.isNumber()');
  } else if (type === 'date') {
    constraints.push('newData.isString()');
    constraints.push(`newData.val().matches(${ISO_8601})`);
  }

  // Other validations
  if (field.isOneOf != null) {
    const choiceConstraints = field.isOneOf.map(o => `newData.val() === ${o}`);
    constraints.push(`(${choiceConstraints.join(' || ')})`);
  }
  if (field.hasAtLeastChars != null) {
    constraints.push(`newData.val().length >= ${field.hasAtLeastChars}`);
  }
  if (field.hasAtMostChars != null) {
    constraints.push(`newData.val().length <= ${field.hasAtMostChars}`);
  }
  if (field.hasLengthWithinRange != null) {
    constraints.push(`newData.val().length >= ${field.hasLengthWithinRange[0]}`);
    constraints.push(`newData.val().length <= ${field.hasLengthWithinRange[1]}`);
  }
  if (field.isEmail) {
    constraints.push('newData.val().matches(/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\\.[A-Z]{2,4}$/i)');
  }
  if (field.isUrl) {
    console.log('WARNING: isUrl validation is currently unsupported');
  }
  if (field.isIp) {
    console.log('WARNING: isIp validation is currently unsupported');
  }
  if (field.isCreditCard) {
    console.log('WARNING: isCreditCard validation is currently unsupported');
  }
  if (field.matchesPattern != null) {
    constraints.push(`newData.val().matches(/${field.matchesPattern[0]}/${field.matchesPattern[1]})`);
  }
  if (field.isGte != null) {
    constraints.push(`newData.val() >= ${field.isGte}`);
  }
  if (field.isLte != null) {
    constraints.push(`newData.val() <= ${field.isLte}`);
  }
  if (field.isWithinRange != null) {
    constraints.push(`newData.val() >= ${field.isWithinRange[0]}`);
    constraints.push(`newData.val() <= ${field.isWithinRange[1]}`);
  }
  if (field.satisfies != null) {
    constraints.push(field.satisfies);
  }
  return constraints;
};

// ====================================
// Helpers
// ====================================
const getRequiredFields = ({ fields, relations }) => {
  const required = [];
  Object.keys(fields).forEach(name => {
    const field = fields[name];
    if (!field.existsInServer) return;
    if (field.isRequired || field.isPrimaryKey) {
      required.push(name);
    }
  });
  Object.keys(relations).forEach(name => {
    const relation = relations[name];
    if (relation.isRequired) required.push(relation.fkName);
  });
  return required;
};

// ====================================
// Public
// ====================================
export { output };
