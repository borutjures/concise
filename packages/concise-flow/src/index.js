// @flow

/* eslint-disable prefer-template */

import fs from 'fs';
import type { Schema, OutputProcessor, SchemaUtils } from 'concise-types';
import upperFirst from 'lodash.upperfirst';

type OutputOptions = {
  file?: string,
};

// ====================================
// Main
// ====================================
const output: OutputProcessor = async (
  schema: Schema,
  options: OutputOptions,
  utils: SchemaUtils,
) => {
  const preprocessedSchema = utils.preprocess(schema);
  const raw = writeTypes(preprocessedSchema);
  if (options.file) fs.writeFileSync(options.file, raw, 'utf8');
  return raw;
};

// ====================================
// Flow writer
// ====================================
const writeTypes = ({ models }) => {
  let out = '';
  Object.keys(models).forEach(modelName => {
    out += writeType(models, modelName);
  });
  return out;
};

const writeType = (models, modelName) => {
  const {
    description,
    fields = {},
    relations = {},
  } = models[modelName];
  const upperModelName = upperFirst(modelName);
  const allSpecs = [];
  Object.keys(fields).forEach(fieldName => {
    allSpecs.push(writeField(fieldName, fields[fieldName]));
  });
  Object.keys(relations).forEach(fieldName => {
    allSpecs.push(writeField(`${fieldName}Id`, relations[fieldName]));
  });
  const contents = allSpecs.length
    ? `\n  ${allSpecs.join('\n  ')}\n`
    : '';
  return '' +
    `// ${upperModelName}\n` +
    (description ? `// ${description}\n` : '') +
    `type ${upperModelName} = {${contents}};\n\n`;
};

const writeField = (name, specs: any) => {
  const required = specs.validations && specs.validations.required ? '' : '?';
  const typeStr = writeFieldType(specs.type);
  const comment = specs.description ? `  // ${specs.description}` : '';
  return `${name}${required}: ${typeStr},${comment}`;
};

const writeFieldType = type => {
  if (type === 'uuid') return 'string';
  if (type === 'json') return 'any';
  if (type === 'date') return 'Date';
  return type;
};

// ====================================
// Public
// ====================================
export { output };
