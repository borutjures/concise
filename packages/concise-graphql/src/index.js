// @flow

/* eslint-disable prefer-template */

import fs from 'fs';
import type { Schema, OutputProcessor, SchemaUtils } from 'concise-types';
import upperFirst from 'lodash.upperfirst';

/* --
Output-only.

Output options:
* `relay?` (`boolean` = `false`): include `Node` interface
  and `node` root field, define connections, etc.
-- */
type OutputOptions = {
  file?: string,
  relay?: boolean,
};

// ====================================
// Main
// ====================================
const output: OutputProcessor = async (
  schema: Schema,
  options: OutputOptions,
  utils: SchemaUtils,
) => {
  const preprocessedSchema = utils.preprocess(schema);
  const raw = writeTypes(preprocessedSchema, options);
  if (options.file) fs.writeFileSync(options.file, raw, 'utf8');
  return raw;
};

// ====================================
// GraphQL writer
// ====================================
const writeTypes = ({ models }, options) => {
  let out = '';
  const relay = options.relay;
  if (relay) {
    out +=
      '# An object with an ID\n' +
      'interface Node {\n' +
      '  # The id of the object\n' +
      '  id: ID!\n' +
      '}\n\n' +
      '# Information about pagination in a connection\n' +
      'type PageInfo {\n' +
      '  # When paginating forwards, are there more items?\n' +
      '  hasNextPage: Boolean!\n' +
      '  # When paginating backwards, are there more items?\n' +
      '  hasPreviousPage: Boolean!\n' +
      '  # When paginating backwards, the cursor to continue\n' +
      '  startCursor: String\n' +
      '  # When paginating forwards, the cursor to continue\n' +
      '  endCursor: String\n' +
      '}\n\n';
  }
  Object.keys(models).forEach(modelName => {
    out += writeType(models, modelName, options);
  });
  const querySpecs = Object.keys(models).map(modelName => {
    const { plural } = models[modelName];
    const typeStr = relay
      ? `${upperFirst(modelName)}Connection`
      : `[${upperFirst(modelName)}]`;
    return `  ${plural}: ${typeStr}\n`;
  });
  if (relay) querySpecs.unshift('  node: Node\n')
  out += (
    '# Root query\n' +
    'type Query {\n' +
    querySpecs.join('') +
    '}\n\n'
  );
  if (relay) out += writeRelayConnections(models);
  return out;
};

const writeType = (models, modelName, options) => {
  const { description, fields = {}, relations = {} } = models[modelName];
  const upperModelName: string = upperFirst(modelName);
  const implementsStr = options.relay ? ' implements Node' : '';
  let allSpecs = [];
  Object.keys(fields).forEach(fieldName => {
    allSpecs = allSpecs.concat(writeField(fieldName, fields[fieldName], options));
  });
  Object.keys(relations).forEach(fieldName => {
    allSpecs = allSpecs.concat(writeField(fieldName, relations[fieldName], options));
  });
  const contents = allSpecs.length ? `\n  ${allSpecs.join('\n  ')}\n` : '';
  let comment = `# ${upperModelName}`;
  if (description) comment += `: ${description}`;
  return (
    `${comment}\n` +
    `type ${upperModelName}${implementsStr} {${contents}}\n\n`
  );
};

const writeField = (name, specs: any, options) => {
  const typeStr = writeFieldType(specs, options);
  const out = [];
  if (specs.description) out.push(`# ${specs.description}`);
  out.push(`${name}: ${typeStr}`);
  return out;
};

const writeFieldType = (specs, options) => {
  const { type, model, isPrimaryKey, validations, isPlural } = specs;
  let out;
  if (isPrimaryKey) out = 'ID';
  else if (model) out = upperFirst(model);
  else if (type === 'boolean') out = 'Boolean';
  else if (type === 'number') out = specs.isFloat ? 'Float' : 'Int';
  else out = 'String';
  if (isPlural) {
    out = options.relay ? `${out}Connection` : `[${out}]`;
  }
  if (isPrimaryKey || (validations && validations.isRequired)) out += '!';
  return out;
};

const writeRelayConnections = models => {
  let out = '';
  const keys = {};
  Object.keys(models).forEach(modelName => {
    const { relations } = models[modelName];
    Object.keys(relations).forEach(relationName => {
      const relation = relations[relationName];
      if (!relation.isPlural || keys[relation.model]) return;
      keys[relation.model] = true;
      const upperModelName: string = upperFirst(relation.model);
      out +=
        `type ${upperModelName}Connection {\n` +
        '  pageInfo: PageInfo!\n' +
        `  edges: [${upperModelName}Edge]\n` +
        '}\n\n' +
        `type ${upperModelName}Edge {\n` +
        `  node: ${upperModelName}\n` +
        '  cursor: String!\n' +
        '}\n\n';
    });
  });
  return out;
};

/*
Ideas for mutations (check Relay Modern docs):

input CompanyUpdateInput {
  id: ID!  // remove this for creates
  ...all company fields as optional
  ...Id fields for (direct-only) relations
  storyId: String  // for use with Storyboard (behind a flag)
  clientMutationId: String
}

type CompanyUpdatePayload {
  company: Company
  clientMutationId: String
}
*/

// ====================================
// Public
// ====================================
export { output };
