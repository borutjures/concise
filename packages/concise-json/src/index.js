// @flow

import fs from 'fs';
import type { Schema, InputProcessor, OutputProcessor } from 'concise-types';

/* --
Input/output.

Output options:
* `file?` (`string`): if specified, output will be written to the specified path
* `prettyJson?` (`boolean` = `false`): prettify JSON output
-- */
type InputOptions = {
  file?: string,
  raw?: string,
};

type OutputOptions = {
  prettyJson?: boolean,
  file?: string,
};

// ====================================
// Processors
// ====================================
const input: InputProcessor = async (options: InputOptions) => {
  let raw: string;
  if (options.raw) {
    raw = options.raw;
  } else if (options.file) {
    raw = fs.readFileSync(options.file, 'utf8');
  } else {
    throw new Error('Specify either `raw` or `file`');
  }
  return JSON.parse(raw);
};

const output: OutputProcessor = async (
  schema: Schema,
  options: OutputOptions
) => {
  const { prettyJson = true } = options;
  const raw = prettyJson
    ? JSON.stringify(schema, null, 2)
    : JSON.stringify(schema);
  if (options.file) fs.writeFileSync(options.file, raw, 'utf8');
  return raw;
};

// ====================================
// Public
// ====================================
export { input, output };
