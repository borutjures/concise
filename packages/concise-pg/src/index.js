// @flow

/* eslint-disable prefer-template */

import fs from 'fs';
import { omit, merge, setIn } from 'timm';
import type { Schema, OutputProcessor } from 'concise-types';

type OutputOptions = {
  file?: string,
};

// ====================================
// Main
// ====================================
const output: OutputProcessor = async (
  schema: Schema,
  options: OutputOptions,
) => {
  const preprocessedSchema = preprocess(schema);
  const raw = schemaToSql(preprocessedSchema, options);
  if (options.file) fs.writeFileSync(options.file, raw, 'utf8');
  return raw;
};

const preprocess = schema => {
  const models = {};

  // Process includes
  Object.keys(schema.models).forEach(modelName => {
    let model = schema.models[modelName];
    if (model.includeOnly) return;
    if (model.includes) {
      Object.keys(model.includes).forEach(includeName => {
        const include = schema.models[includeName];
        if (!include) throw new Error(`INCLUDE_NOT_FOUND ${modelName}/${includeName}`);
        const { fields, relations } = include;
        model = merge(omit(model, ['includes']), {
          fields: fields ? merge(fields, model.fields) : undefined,
          relations: relations ? merge(relations, model.relations) : undefined,
        });
      });
    }
    models[modelName] = model;
  });

  // Process relation field types
  Object.keys(models).forEach(modelName => {
    const model = models[modelName];
    Object.keys(model.relations || {}).forEach(relationName => {
      const relation = model.relations[relationName];
      const relatedModel = models[relation.model];
      if (!relatedModel) {
        throw new Error(`RELATED_MODEL_NOT_FOUND ${modelName}/${relationName}/${relation.model}`);
      }
      const idField = relatedModel && relatedModel.fields && relatedModel.fields.id;
      if (!idField) {
        throw new Error(`ID_FIELD_NOT_FOUND ${modelName}/${relationName}/${relation.model}`);
      }
      models[modelName] = setIn(
        models[modelName],
        ['relations', relationName, 'type'],
        idField.type
      );
    });
  });

  return { models };
};

const schemaToSql = ({ models }, options) => {
  let out = '';
  Object.keys(models).forEach(modelName => {
    out += sqlTable(models, modelName, options);
  });
  return out;
};

const sqlTable = (models, modelName, options) => {
  let sqlFields = [];
  let sqlConstraints = [];
  const sqlComments = [];
  let tableName = `"${modelName}"`;
  if (options.schema) tableName = `"${options.schema}".${tableName}`;
  const {
    description,
    fields = {},
    relations = {},
  } = models[modelName];
  if (description) {
    sqlComments.push(writeComment('TABLE', tableName, description));
  }
  Object.keys(fields).forEach(fieldName => {
    const { sqlField, sqlFieldComment, sqlFieldConstraints } =
      writeField(models, modelName, tableName, fieldName, options);
    if (sqlField) sqlFields.push(sqlField);
    if (sqlFieldComment) sqlComments.push(sqlFieldComment);
    if (sqlFieldConstraints) sqlConstraints = sqlConstraints.concat(sqlFieldConstraints);
  });
  Object.keys(relations).forEach(relationName => {
    const { sqlFields: someSqlFields, sqlFieldComment } =
      writeForeignKey(models, modelName, tableName, relationName, options);
    if (someSqlFields) sqlFields = sqlFields.concat(someSqlFields);
    if (sqlFieldComment) sqlComments.push(sqlFieldComment);
  });
  const sqlTableEntries = sqlFields.concat(sqlConstraints);
  return `CREATE TABLE ${tableName} (\n` +
    `${sqlTableEntries.map(o => `  ${o}`).join(',\n')}\n` +
    ');\n' +
    sqlComments.map(o => `${o}\n`).join('') +
    '\n';
};

const writeField = (models, modelName, tableName, fieldName) => {
  const field = models[modelName].fields[fieldName];
  const segments = [`"${fieldName}"`, writeFieldType(field)];
  if (field.validations && field.validations.required) {
    segments.push('NOT NULL');
  }
  const { default: defVal } = field;
  if (defVal != null) {
    if (defVal === true || defVal === false) {
      segments.push(`DEFAULT ${defVal}`);
    } else {
      segments.push(`DEFAULT '${defVal}'`);
    }
  }
  const sqlField = segments.join(' ');
  const sqlFieldComment = field.description
    ? writeComment('FIELD', `${tableName}."${fieldName}"`, field.description)
    : undefined;
  const sqlFieldConstraints = [];
  if (field.primaryKey) {
    sqlFieldConstraints.push(`CONSTRAINT "${modelName}_pk" PRIMARY KEY (${fieldName})`);
  }
  return { sqlField, sqlFieldComment, sqlFieldConstraints };
};

const writeForeignKey = (models, modelName, tableName, relationName) => {
  const relation = models[modelName].relations[relationName];
  const fieldName = `${relationName}Id`;
  const sqlFields = [];
  const required = relation.validations && relation.validations.required;

  // Foreign key
  let segments;
  segments = [`"${fieldName}"`, writeFieldType(relation)];
  if (required) segments.push('NOT NULL');
  sqlFields.push(segments.join(' '));

  // For polymorphic relationships, additional qualifier
  if (relation.polymorphic) {
    segments = [`"${relationName}Type"`, writeFieldType({ type: 'string' })];
    if (required) segments.push('NOT NULL');
    sqlFields.push(segments.join(' '));
  }

  const sqlFieldComment = relation.description
    ? writeComment('FIELD', `${tableName}."${fieldName}"`, relation.description)
    : undefined;
  return { sqlFields, sqlFieldComment };
};

const writeFieldType = field => {
  const { type } = field;
  if (type === 'string') return field.long ? 'text' : 'character varying(255)';
  if (type === 'number') return field.float ? 'double precision' : 'integer';
  if (type === 'date') return 'timestamp with time zone';
  if (type === 'json') return 'jsonb';
  return field.type;
};

const writeComment = (type, object, cmt) =>
  `COMMENT ON ${type} ${object} IS '${cmt}';`;

// ====================================
// Public
// ====================================
export { output };
