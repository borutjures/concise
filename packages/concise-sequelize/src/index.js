// @flow

/* eslint-disable prefer-template */

import type { Schema, OutputProcessor, SchemaUtils, Field } from 'concise-types';
import upperFirst from 'lodash.upperfirst';

/* --
Output-only.

Output options:
* `sequelize` (`Object`)
-- */
type OutputOptions = {
  sequelize: Object,
};

// ====================================
// Main
// ====================================
const output: OutputProcessor = async (
  schema: Schema,
  options: OutputOptions,
  utils: SchemaUtils,
) => {
  const preprocessedSchema = utils.preprocess(schema);
  const { sequelize } = options;
  const db = { sequelize };
  defineModels(db, preprocessedSchema);
  defineRelations(db, preprocessedSchema);
  return db;
};

/* eslint-disable no-param-reassign */
const defineModels = (db, schema) => {
  const { models } = schema;
  Object.keys(models).forEach(modelName => {
    console.log(`${modelName}`)
    const model = models[modelName];
    const className = upperFirst(modelName);
    db[className] = db.sequelize.define(modelName, defineFields(db, model), {
      validate: {},
      classMethods: {},
      instanceMethods: {},
    });
  });
};

const defineRelations = (db, schema) => {};
/* eslint-enable no-param-reassign */

// ====================================
// Spec builders
// ====================================
const defineFields = (db, model) => {
  const { fields } = model;
  const out = {};
  Object.keys(fields).forEach(fieldName => {
    const field = fields[fieldName];
    const newField = {};
    const { isPrimaryKey, isAutoIncrement, defaultValue } = field;
    const sequelizeType = getFieldType(field);
    newField.type = db.sequelize[sequelizeType];
    if (isPrimaryKey != null) newField.primaryKey = isPrimaryKey;
    if (isAutoIncrement != null) newField.autoIncrement = isAutoIncrement;
    if (defaultValue !== undefined) newField.defaultValue = defaultValue;
    console.log(` - ${fieldName}: ${sequelizeType}`)
    out[fieldName] = newField;
  });
  return out;
};

const getFieldType = (field: Field) => {
  const { type } = field;
  if (type === 'string') return field.isLong ? 'TEXT' : 'STRING';
  if (type === 'boolean') return 'BOOLEAN';
  if (type === 'uuid') return 'STRING';
  if (type === 'json') return 'JSONB';
  if (type === 'number') return field.isFloat ? 'DOUBLE' : 'INTEGER';
  if (type === 'date') {
    if (field.noDate) return 'TIME';
    if (field.noTime) return 'DATEONLY';
    return 'DATE';
  }
  throw new Error(`UNSUPPORTED_FIELD_TYPE ${type}`);
};

// ====================================
// Public
// ====================================
export { output };
