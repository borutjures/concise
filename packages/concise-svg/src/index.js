// @flow

/* eslint-disable prefer-template */

import fs from 'fs';
import Viz from 'viz.js';
import { addDefaults } from 'timm';
import type {
  Schema,
  OutputProcessor,
  SchemaUtils,
  ModelName,
  FieldName,
} from 'concise-types';

type OutputOptions = {
  file?: string,
  filterEdges?: (
    { from: ModelName, to: ModelName, as: FieldName, required: boolean },
  ) => boolean,
  edgeLabels?: boolean,
};

const DEFAULT_OPTIONS = {
  edgeLabels: true,
};

// ====================================
// Main
// ====================================
const output: OutputProcessor = async (
  schema: Schema,
  options: OutputOptions,
  utils: SchemaUtils,
) => {
  const preprocessedSchema = utils.preprocess(schema);
  const vizInput = writeDiagram(
    preprocessedSchema,
    addDefaults(options, DEFAULT_OPTIONS),
  );
  const svg = Viz(vizInput);
  if (options.file) {
    fs.writeFileSync(options.file, svg, 'utf8');
  }
  return svg;
};

// ====================================
// Flow writer
// ====================================
const writeDiagram = ({ models }, { filterEdges, edgeLabels }) => {
  // Object.keys(models).forEach(modelName => {
  //   out += writeType(models, modelName);
  // });
  const modelNames = Object.keys(models);
  const nodes = [];
  modelNames.forEach(modelName => {
    // const spec = models[modelName];
    // const fieldNames = ['id'].concat(Object.keys(spec.relations || {})).map(o => `<${o}> ${o}`);
    // nodes.push(`${modelName} [label="{${modelName} | ${fieldNames.join('|')}}"]`);
    nodes.push(modelName);
  });
  const edges = [];
  modelNames.forEach(modelName => {
    const { relations } = models[modelName];
    Object.keys(relations || {}).forEach(fieldName => {
      const specs = relations[fieldName];
      const required = specs.validations && specs.validations.required === true;
      if (
        filterEdges &&
        !filterEdges({
          from: modelName,
          to: specs.model,
          as: fieldName,
          required,
        })
      ) {
        return;
      }
      // let edge = `${modelName}:${fieldName} -> ${specs.model}:id [label="${fieldName}"]`;
      const edgeLabel = edgeLabels && fieldName !== specs.model
        ? ` [label="${fieldName}"]`
        : '';
      let edge = `${modelName} -> ${specs.model}${edgeLabel}`;
      if (!required) edge += ' [style=dotted]';
      edges.push(edge);
    });
  });
  return 'digraph {\n' +
    '  node [shape=box];\n' +
    nodes.map(o => `  ${o};\n`).join('') +
    edges.map(o => `  ${o};\n`).join('') +
    '}\n';
};

// ====================================
// Public
// ====================================
export { output };
