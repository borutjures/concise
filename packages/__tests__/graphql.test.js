/* eslint-env jest */

import path from 'path';
import Concise from '../concise/src';
import { input as yamlInput } from '../concise-yaml/src';
import { output } from '../concise-graphql/src';

describe('GraphQL schema output', () => {
  it('schema1.yaml -> GraphQL', async () => {
    const concise = new Concise();
    await concise.input(yamlInput, {
      file: path.join(__dirname, './fixtures/schema1.yaml'),
    });
    const gqlSchema = await concise.output(output);
    expect(gqlSchema).toMatchSnapshot();
  });

  it('schema1.yaml + schema2.yaml -> GraphQL', async () => {
    const concise = new Concise();
    await concise.input(yamlInput, {
      file: path.join(__dirname, './fixtures/schema1.yaml'),
    });
    await concise.input(yamlInput, {
      file: path.join(__dirname, './fixtures/schema2.yaml'),
    });
    const gqlSchema = await concise.output(output);
    expect(gqlSchema).toMatchSnapshot();
  });
});
