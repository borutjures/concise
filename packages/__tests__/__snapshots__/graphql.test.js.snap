// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL schema output schema1.yaml + schema2.yaml -> GraphQL 1`] = `
"# User: A user of the platform
type User {
  id: ID!
  notes: String
  # Name of the user
  name: String!
  surname: String
  aBoolean: Boolean!
  projects: [Project]
  createdProjects: [Project]
}

# Project
type Project {
  id: ID!
  notes: String
  name: String
  startDate: String
  endDate: String
  projectManager: User!
  technicalManager: User
  creator: User
}

# Root query
type Query {
  users: [User]
  projects: [Project]
}

"
`;

exports[`GraphQL schema output schema1.yaml -> GraphQL 1`] = `
"# User: A user of the platform
type User {
  id: ID!
  notes: String
  # Name of the user
  name: String!
  surname: String
  aBoolean: Boolean!
  projects: [Project]
}

# Project
type Project {
  id: ID!
  notes: String
  name: String
  projectManager: User!
  technicalManager: User
}

# Root query
type Query {
  users: [User]
  projects: [Project]
}

"
`;
